### ISSUE TITLE
ops defined inside tf.while_loop's cond/body or tf.cond's true_fn/false_fn functions ignore their enclosed tf.device if the tf.while_loop/tf.cond itself is inside a tf.device

### ISSUE THREAD

**Issue by dhruvin2910** [tensorflow/tensorflow#44011]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Catalina
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: Python 3.7.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

`tf.print` doesn't print on specified device (using `tf.device`) if it is inside a for loop in graph mode.

Initially I though it's an autograph issue. But if you trace the graph and inspect it in tensorboard (color by device), the graph does have that PrintV2 placed where it's supposed to.

Now, I'm not sure why it's happening.
I'm yet to verify if this happens across all the ops, or just the `tf.print`.

Waiting for the team to let me know what to check next.

**Describe the expected behavior**

`tf.print` should print on specified device (using `tf.device`). ~Everything works fine if I'm using tf.while_loop instead.
Can be seen from sample code.~

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**THIS SNIPPET IS NOT CORRECT ANYMORE, PLEASE READ A COUPLE OF COMMENTS BELOW.**

```python
import tensorflow as tf

# assume there's a tf.distribute.Server running on rpi.local:2222
tf.config.experimental_connect_to_cluster(
    tf.train.ClusterSpec({'worker': ['dhruvins-macbook-air.local:2222', 'rpi.local:2222']}),
    job_name='worker',
    task_index=0
)

laptop = '/job:worker/task:0'
rpi = '/job:worker/task:1'


@tf.function
def test_for():
    with tf.device(rpi):
        for i in tf.range(3):
            tf.print('rpi', 'for', i)
            with tf.device(laptop):
                tf.print('laptop', 'for', i)


@tf.function
def test_while():
    def cond(i):
        return i < 3

    def body(i):
        with tf.device(rpi):
            tf.print('rpi', 'while', i)
            with tf.device(laptop):
                tf.print('laptop', 'while', i)
        return [i + 1]

    tf.while_loop(cond, body, [0], parallel_iterations=1)


test_for()
test_while()

```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@jvishnuvardhan 
I ran this code on nightly, the code keep running and does not produce any output, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/cdad7be80aca26f332c5e366cfbced34/untitled445.ipynb).

**Comment by dhruvin2910** [comment]
@Saduf2019 One will need to modify their colab code a bit to point to existing (running) `tf.distribute.Server`(s).
The reason why it's stuck is because it's unable to connect to rpi.local:2222 while running inside colab's container.
The sample code I provided requires at least two tensorflow servers. On same machine, or on different ones (i.e. my case).
I'll soon share a link to a colab notebook reproducing this issue.

**Comment by mdanatg** [comment]
@dhananjay92 there is a notable difference between the two snippets of code: In the autograph case, the `tf.device` is places outside the loop. In the tf.while_loop case, the `tf.device` is placed inside the loop. So to make them equivalent, you'd need to either write:

```
for i in tf.range(3):
    with tf.device(rpi):
```

or 

```
with tf.device(rpi):
    def cond(i):
        ...
    def body(i):
        ...

    tf.while_loop(...)
```

And I believe that highlights the bug: if we place a `tf.device` outside a `tf.while_loop`, its body does not respect it. Things work as expected when the `tf.device` is placed inside the loop body. Could you verify that in your setup?

At any rate, this looks like a placement bug.

**Comment by JayNakrani** [comment]
@mdanatg You sure this is intended for me? Did you mean to tag dhruvin2910 and accidentally selected me in dropdown somehow? Because although I have contributed to tensorflow in past, I don't think I have touched this part of the code.

**Comment by mdanatg** [comment]
@JayNakrani sorry about that. I meant to tag @dhruvin2910. Not sure how the autocomplete filled your tag.

**Comment by dhruvin2910** [comment]
> there is a notable difference between the two snippets of code: In the autograph case, the tf.device is places outside the loop. In the tf.while_loop case, the tf.device is placed inside the loop. So to make them equivalent, you'd need to either write: ...

@mdanatg Yes, I intended to provide what you just suggested. The problem originated from a bit involving code. Somehow, I missed wrapping the `while_loop` with `tf.device` instead, while I was simplifying it to reproduce this issue.

> And I believe that highlights the bug: if we place a tf.device outside a tf.while_loop, its body does not respect it. Things work as expected when the tf.device is placed inside the loop body. Could you verify that in your setup?

Yes. [Updated Colab](https://colab.research.google.com/gist/dhruvin2910/d8b0765a28b8182303105b29dd57262c/untitled445.ipynb#scrollTo=vFcc6P_S_ebz)

**Comment by dhruvin2910** [comment]
As I mentioned earlier, the graph seems to have `PrintV2` placed on expected device.

I'm unable to upload my logs properly to tensorboard.dev, so here's a screenshot (blue ones are on rpi and green ones are on my macbook):

![The Graph](https://user-images.githubusercontent.com/9679326/96669104-27d2db00-137a-11eb-8dc9-63eeaae26a48.png)

**Comment by dhruvin2910** [comment]
I tested if all ops are misplaced or not with this code.

I don't really know a proper way to find it out. But here's an experiment (and assumptions) to test it:

1. A very long while loop may be substituted for a long running task (with parallel iterations set to 1)
2.  1. With proper `control_dependencies` set, one can measure computation time of a set of ops.
     2. If two devices have different compute power, one can guess the device placement by the amount of time taken, faster task would mean it was placed on the device with better compute.
3. 1. Since both long running tasks are assigned to different devices, they should run concurrently.
    2. One can guess the device placement is wrong if they are run in sequence. Can be guessed from total time, or from print statements.

```python

def task(n):
    with tf.name_scope('task'):
        return tf.while_loop(lambda i: i < n, lambda i: i + 1, [0], parallel_iterations=1)[0]


def timed(f, *args):
    with tf.name_scope('timed'):
        start = tf.timestamp(name='start')
        with tf.control_dependencies([start]):
            r = f(*args)
        with tf.control_dependencies([r]):
            end = tf.timestamp(name='end')
        return tf.subtract(end, start, name='time'), r


@tf.function
def test_while():
    with tf.device(rpi):
        def cond(i):
            return i < 1

        def body(i):
            t, n = timed(task, 100_000)
            tf.print('rpi', 'while', t, n)
            with tf.device(laptop):
                t, n = timed(task, 100_000)
                tf.print('laptop', 'while', t, n)
            return [i + 1]

        tf.while_loop(cond, body, [0], parallel_iterations=1)
```

Observations:
1. The tasks are running in sequence, so there's something more to this story.
2. I have a rasppberry pi running at 700MHz and a MacBook Air running at 1.8GHz. The tasks take roughly the same time (6s), and the total time was around 12s. Again there's something missing.
```
rpi while 6.2778100967407227 100000
laptop while 6.16434907913208 100000
```

Note:
Since I may have made mistakes in my assumptions or the code, I don't think the experiment is conclusive.
I think it may still help further investigation.

**Comment by dhruvin2910** [comment]
Just discovered `tf.debugging.set_log_device_placement(True)`, and I can confirm that the device placement is indeed incorrect.

```
...
while/body/_1/while/PrintV2: (PrintV2): /job:worker/replica:0/task:1/device:CPU:0
...
while/body/_1/while/PrintV2_1: (PrintV2): /job:worker/replica:0/task:1/device:CPU:0
...
```

Both of these are placed on `task:1` (aka `rpi`).
Also, not just the `tf.print`, but all the ops in the loop are are placed incorrectly. Tested with `tf.random.uniform`.

**Comment by dhruvin2910** [comment]
`tf.cond` and `if ...: else: ...` (with autograph) both have same problem.

```python

@tf.function
def test_cond():
    with tf.device(rpi):
        def true_fn():
            tf.print('rpi')
            with tf.device(laptop):
                tf.print('laptop')

        def false_fn():
            tf.print('rpi')
            with tf.device(laptop):
                tf.print('laptop')

        tf.cond(tf.greater_equal(tf.random.uniform(()), 0.5), true_fn, false_fn)


@tf.function
def test_if():
    with tf.device(rpi):
        if tf.greater_equal(tf.random.uniform(()), 0.5):
            tf.print('rpi')
            with tf.device(laptop):
                tf.print('laptop')
        else:
            tf.print('rpi')
            with tf.device(laptop):
                tf.print('laptop')
```

No `tf.print` (or any other op placed via `tf.device`) will be placed on laptop

**Comment by dhruvin2910** [comment]
After a lot of experiments, here's a concise example of what is working and what is not.

```python
import tensorflow as tf

# assume there's a tf.distribute.Server running on rpi.local:2222
tf.config.experimental_connect_to_cluster(
    tf.train.ClusterSpec({'worker': ['dhruvins-macbook-air.local:2222', 'rpi.local:2222']}),
    job_name='worker',
    task_index=0
)

laptop = '/job:worker/task:0'
rpi = '/job:worker/task:1'


@tf.function
def works():
    def true_fn():
        # this tf.device will be respected
        with tf.device(laptop):
            tf.print('expected on', 'laptop')

    def false_fn():
        # this tf.device will be respected
        with tf.device(rpi):
            tf.print('expected on', 'rpi')

    tf.cond(tf.constant(True), true_fn, false_fn)


@tf.function
def does_not_work():
    # we enclose everything inside a tf.device
    with tf.device(rpi):
        def true_fn():
            # this tf.device will be ignored
            with tf.device(laptop):
                tf.print('expected on', 'laptop')

        def false_fn():
            # this tf.device will be ignored
            with tf.device(rpi):
                tf.print('expected on', 'rpi')

        # tf.cond is in tf.device now
        tf.cond(tf.constant(True), true_fn, false_fn)


works()
does_not_work()
```

Both calls must print 'expected on laptop' on laptop.
But the second call (aka `does_not_work`) prints on rpi instead.

@mdanatg can you verify this?

**Comment by mdanatg** [comment]
Thanks for the detailed investigation. We're having a closer look at the cause.

Just to double check, in the last snippet you meant `tf.device(laptop)`, right (since the outer device is already `rpi`)?

**Comment by dhruvin2910** [comment]
> Just to double check, in the last snippet you meant tf.device(laptop), right (since the outer device is already rpi)?

`works` used to flip a coin and printed on a device at random, but to make the code deterministic, I set the condition as `True`. So in the example `false_fn` is never evaluated and any of `rpi` or `laptop` would work.

I tried to demonstrate that if you take a correctly behaving `tf.cond` code (`works` in our example), and wrap it with a `tf.device`, the ops inside `true_fn` and `false_fn` don't get distributed the way they were originally.

All the ops are placed where `tf.cond` is placed.

**Comment by allenlavoie** [comment]
https://github.com/tensorflow/tensorflow/commit/d3698cdfd94c858092ebbb9af7ab0815e9bc78c1 should fix the `tf.device` scope inheritance issue. It also has a test that ops don't always follow cond placement (even if there's a device scope around the cond).

Please give it a try and re-open if something's still broken.

**Comment by dhruvin2910** [comment]
Thank you so much @allenlavoie. I'll see if the issue is fixed in the nightly (when the commit gets in nightly) and let you know.

For my future reference, [colab](https://colab.research.google.com/gist/dhruvin2910/d8b0765a28b8182303105b29dd57262c/untitled445.ipynb#scrollTo=vFcc6P_S_ebz)

**Comment by dhruvin2910** [comment]
I think the commit you added must have arrived in nightly.
I have been trying to see if the issue is still there or not for 3 days, and the problem persists.
See the colab above.

PS: I'm unable to reopen this issue.

**Comment by mdanatg** [comment]
@dhruvin2910 I'm not sure if the issue can be accurately reproduced in the colab setup. IIUC, the colab doesn't see the output from any of the two cluster workers. At the same time, I suspect w1_proc will capture outputs from both workers.

Looking at the op placement: `print(test_while.get_concrete_function().graph.as_graph_def())`, if we search for `op: "PrintV2"`, I can see that the two Print ops are placed one on `worker/task:0` and `worker/task:1`, as expected.

I don't know if it's possible to configure a cluster that includes the current colab process. @guptapriya ?

**Comment by dhruvin2910** [comment]
> I'm not sure if the issue can be accurately reproduced in the colab setup. IIUC, the colab doesn't see the output from any of the two cluster workers. At the same time, I suspect w1_proc will capture outputs from both workers.

@mdanatg I tried a simple `tf.function` code (example below) and the workers had correct outputs.

```python
with tf.device(w0):
      tf.print("This is a test print. Should be done on w0")
with tf.device(w1):
      tf.print("This is a test print. Should be done on w1")
```

> Looking at the op placement: print(test_while.get_concrete_function().graph.as_graph_def()), if we search for op: "PrintV2", I can see that the two Print ops are placed one on worker/task:0 and worker/task:1, as expected.

I mentioned exactly this in my previous https://github.com/tensorflow/tensorflow/issues/44011#issuecomment-713271763. Ops in the graph have proper device assigned, but the ops are placed incorrectly at runtime. See this https://github.com/tensorflow/tensorflow/issues/44011#issuecomment-713360321.

**Comment by mdanatg** [comment]
Thanks, and sorry I missed those bits. I concur, there is a bug still present in the placement logic. I suggest updating the test colab to make that clearer:

```
@tf.function
def test_while():
    with tf.device(w0):
      tf.print('w0', 'control test')
    with tf.device(w1):
      ...
```

**Comment by allenlavoie** [comment]
The second issue looks like it was related to function inlining (when inlining, the function's placement partially overrode the body's placement; conds were inlined as two functions). I have a change that should go through soon for that.

Thank you for the report.

**Comment by dhruvin2910** [comment]
The issue seems to be fixed in 2.4 🎉. Thanks @mdanatg and @allenlavoie.
